src/Clapi/Serialisation.hs

```diff
@@ -139,53 +146,196 @@ cvBuilder (ClFloat x) = return $ floatBE x
+    p te
+
+tdTotalBuilder :: TaggedData e a -> (a -> CanFail Builder) -> a -> CanFail Builder
+tdTotalBuilder td bdr a = builder (tdInstanceToTag td $ a) <<>> bdr a
+
+instance Serialisable UMsgError where
+    builder (UMsgError p s) = builder p <<>> builder s
     parser = do
-        len <- parser :: Parser Word16
-        messages <- count (fromIntegral len) (parser :: Parser Message)
-        return messages
+        p <- parser
+        s <- parser
+        return $ UMsgError p s
+
+data SubMsgType
+  = SubMsgTSub
+  | SubMsgTUnsub deriving (Enum, Bounded)
+
+subMsgTaggedData = taggedData typeToTag msgToType
+  where
+    typeToTag (SubMsgTSub) = 'S'
+    typeToTag (SubMsgTUnsub) = 'U'
+    msgToType (UMsgSubscribe _) = SubMsgTSub
+    msgToType (UMsgUnsubscribe _) = SubMsgTUnsub
+
+instance Serialisable SubMessage where
+    builder = tdTotalBuilder subMsgTaggedData $ builder . subMsgPath
+    parser = tdTotalParser subMsgTaggedData $ \e -> case e of
+        (SubMsgTSub) -> UMsgSubscribe <$> parser
+        (SubMsgTUnsub) -> UMsgUnsubscribe <$> parser
+
+data DataUpdateMsgType
+  = DUMTAdd
+  | DUMTSet
+  | DUMTRemove
+  | DUMTClear
+  | DUMTSetChildren deriving (Enum, Bounded)
+
+dumtTaggedData = taggedData typeToTag msgToType
+  where
+    typeToTag (DUMTAdd) = 'a'
+    typeToTag (DUMTSet) = 's'
+    typeToTag (DUMTRemove) = 'r'
+    typeToTag (DUMTClear) = 'c'
+    typeToTag (DUMTSetChildren) = 'C'
+    msgToType (UMsgAdd _ _ _ _ _ _) = DUMTAdd
+    msgToType (UMsgSet _ _ _ _ _ _) = DUMTSet
+    msgToType (UMsgRemove _ _ _ _) = DUMTRemove
+    msgToType (UMsgClear _ _ _ _) = DUMTClear
+    msgToType (UMsgSetChildren _ _ _) = DUMTSetChildren
+
+dumtParser e = case e of
+    (DUMTAdd) -> sap UMsgAdd
+    (DUMTSet) -> sap UMsgSet
+    (DUMTRemove) -> rcp UMsgRemove
+    (DUMTClear) -> rcp UMsgClear
+    (DUMTSetChildren) -> do
+        p <- parser
+        ns <- parseListN
+        a <- parser
+        return $ UMsgSetChildren p ns a
+  where
+    sap mt = do
+        p <- parser
+        t <- parser
+        v <- parser
+        i <- parser
+        a <- parser
+        s <- parser
+        return $ mt p t v i a s
+    rcp mt = do
+        p <- parser
+        t <- parser
+        a <- parser
+        s <- parser
+        return $ mt p t a s
+
+dumtBuilder m = case m of
+    (UMsgAdd p t v i a s) -> builder p <<>> builder t <<>> builder v <<>> builder i <<>> builder a <<>> builder s
+    (UMsgSet p t v i a s) -> builder p <<>> builder t <<>> builder v <<>> builder i <<>> builder a <<>> builder s
+    (UMsgRemove p t a s) -> builder p <<>> builder t <<>> builder a <<>> builder s
+    (UMsgClear p t a s) -> builder p <<>> builder t <<>> builder a <<>> builder s
+    (UMsgSetChildren p ns a) -> builder p <<>> encodeListN ns <<>> builder a
+
+instance Serialisable DataUpdateMessage where
+    builder = tdTotalBuilder dumtTaggedData dumtBuilder
+    parser = tdTotalParser dumtTaggedData dumtParser
+
+data TUMT
```

Comment:
Happy for the values to have a mildly cryptic name, but I've already forgotten what this stood for! `TreeUpdateMessageType`?

Reply:
That's the one, yeah, the names stink.